### PR TITLE
[QEff.finetuning] Adding sample config and ReadMe file

### PR DESCRIPTION
**Pull request by tchawada**
Added Readme file for the parameters used in sample config.

### CODE REVIEW

File: QEfficient/finetune/experimental/configs/default_config.yaml [hunk 1]
```diff
@@ -95,10 +95,10 @@ optimizers:
 scheduler:
   scheduler_name: "cosine"
   warmup_steps: 100   # warmup_steps or warmup_ratio
+  warmup_ratio: 0.1
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/default_config.yaml]: I think only one of the warmup_steps or warmup_ratio is available in latest HF. Please remove the other, remove the comment and update accordingly in the readme.

File: QEfficient/finetune/experimental/configs/README.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    *   **warmup\_steps**: Number of steps or ratio (e.g., `100` steps or `0.05` for 5% of total steps).
+    *   Stabilizes early training and improves convergence.
+
+***
+
+### 5. Callbacks
+
+Callbacks allow custom actions during training, such as logging, early stopping, or hardware profiling.
+
+*   **early\_stopping**: Stops training if no improvement in a monitored metric for a defined patience period.
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/README.md]: Add the options available for each callback and its description as well.

File: QEfficient/finetune/experimental/configs/README.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+*   **scheduler**: `cosine` – Learning rate decay strategy; options: `linear`, `cosine`, `cosine_with_restarts`, `polynomial`, `constant`, `constant_with_warmup`, `inverse_sqrt`.
+    *   **warmup\_steps**: Number of steps or ratio (e.g., `100` steps or `0.05` for 5% of total steps).
+    *   Stabilizes early training and improves convergence.
+
+***
+
+### 5. Callbacks
+
+Callbacks allow custom actions during training, such as logging, early stopping, or hardware profiling.
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/README.md]: Add a little detailed description. e.g. Once these callbacks are registered, the trainer class will call these callbacks based on the state of the training. If a callback has "on_epoch_end" method defined then this method will be executed at the end of each epoch.

You can also add some reference to the HF's callback infra to make it more understandable: https://github.com/huggingface/transformers/blob/a7f29523361b2cc12e51c1f5133d95f122f6f45c/src/transformers/trainer_callback.py#L295

https://huggingface.co/docs/transformers/en/main_classes/callback

File: QEfficient/finetune/experimental/configs/README.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+*   **torch\_compile**: Wraps your model with torch.compile() (PyTorch 2.0+) to fuse ops, reduce Python overhead, and generate optimized kernels—often yielding speed-ups without code changes.
+*   **Optional distributed configs**: FSDP, DeepSpeed, or DDP for multi-QAIC or large-scale training.
+*    **resume_from_checkpoint**: Path to a checkpoint to resume training from.
+*    **disable_tqdm**: False by default; set to True to disable progress bar (if running in Notebook).
+
+***
+
+### 4. Optimizer & Scheduler
+
+*   **optimizer**: `adamw` – Optimizer for weight-decoupled regularization; options: `adamw`, `adam`, `sgd`.
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/README.md]: Here we are referring lower case but registry has happened differently. It has "AdamW" instead of adamw.

https://github.com/quic/efficient-transformers/blob/ft_experimental/QEfficient/finetune/experimental/core/optimizer.py#L16C21-L16C25

I think it is better to keep it lower case in config and update the registry code to lower case as well.

File: QEfficient/finetune/experimental/configs/README.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+This section defines core parameters for fine-tuning and evaluation.
+
+*   **type**: `sft` → Specifies training type; `sft` means Supervised Fine-Tuning.
+*   **output\_dir**: Directory where model checkpoints and logs are saved.
+*   **do\_eval**: Enables evaluation during training.
+*   **eval\_strategy**: `epoch` → When to run evaluation (e.g., per epoch or steps).
+*   **gradient\_accumulation\_steps**: Accumulate gradients over multiple steps to simulate larger batch size.
+*   **dtype**: `fp16` → Mixed precision for faster training and reduced memory usage.
+*   **gradient\_checkpointing**: Saves memory by recomputing activations during backward pass (slower but memory-efficient).
+*   **torch\_compile**: Wraps your model with torch.compile() (PyTorch 2.0+) to fuse ops, reduce Python overhead, and generate optimized kernels—often yielding speed-ups without code changes.
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/README.md]: Should we include the default values of these arguments as well here?

File: QEfficient/finetune/experimental/configs/README.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+  
+```
+
+
+***
+
+### 3. Training Configuration
+
+This section defines core parameters for fine-tuning and evaluation.
+
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/README.md]: Based on our last discussion, we thought of skipping some easy to understand options. But now I think it is better to include all the options in the ReadMe file. Because those options do exist in code but no documentation is there to back it with description. If it is there in code then we should properly document it.

Sorry if this sounds opposite to what we had discussed. Let me know your thoughts on this.

The options such as:
- save_total_limit
- metric_for_best_model
- do_eval
- num_train_epochs
- max_steps
- log_level
- log_on_each_node
- logging_strategy
- logging_steps
- save_strategy

File: QEfficient/finetune/experimental/configs/README.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+*   **type**: `sft` → Specifies training type; `sft` means Supervised Fine-Tuning.
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/README.md]: Add some description that using "sft" will use trl's SFTTrainer infrastructure to perform PEFT based SFT training. "base" will use transformers' Trainer infrastructure. If user has written and registered some custom trainer then the same can be called by mentioning the registration key name here.

File: QEfficient/finetune/experimental/configs/README.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+***
+
+
+### 2. Dataset Configuration
+
+This section defines parameters for dataset handling during fine-tuning with Hugging Face models. It covers dataset type, splits, prompt formatting, and DataLoader settings.
+
+*   **tokenizer\_name**: Matches model name.
+*   **dataset\_type**: `seq_completion` → Used for sequence continuation tasks, where the model     predicts the next tokens given an input text (e.g., summarization, text generation).
+*   **dataset\_name**: Dataset name for training.
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/README.md]: Check how we can handle json file here?
https://github.com/quic/efficient-transformers/blob/866a14004258f6cb59a5c63b18786ee8ccadb851/QEfficient/finetune/experimental/core/dataset.py#L62

May be we need to introduce json_file_path as well in config, config manager, its test case and its documentation. 
The json_file_path and dataset_name both are mutually exclusive.

File: QEfficient/finetune/experimental/configs/README.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+*   **train\_split/test\_split**: Defines splits.
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/README.md]: Names of train and test splits to be used in case of dataset being loaded from Huggingface using dataset_name argument.

File: QEfficient/finetune/experimental/configs/README.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+*   **split\_ratio**: For spliting the train/test dataset, only if train split is provided.
+*   **prompt\_func**: Python function to format prompts.
+*   **completion\_template**: `{output}` → string pattern that tells the fine-tuning pipeline which part of the dataset should be treated as the target output (completion) for the model to learn.
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/README.md]: Add 
max_length
group_by_length 
params.

I think other params can also be added as per my below comment for training section. Let me know your thoughts.

File: QEfficient/finetune/experimental/configs/README.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+### 2. Dataset Configuration
+
+This section defines parameters for dataset handling during fine-tuning with Hugging Face models. It covers dataset type, splits, prompt formatting, and DataLoader settings.
+
+*   **tokenizer\_name**: Matches model name.
+*   **dataset\_type**: `seq_completion` → Used for sequence continuation tasks, where the model     predicts the next tokens given an input text (e.g., summarization, text generation).
+*   **dataset\_name**: Dataset name for training.
+*   **train\_split/test\_split**: Defines splits.
+*   **split\_ratio**: For spliting the train/test dataset, only if train split is provided.
+*   **prompt\_func**: Python function to format prompts.
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/README.md]: Add some description about prompt_func and prompt_template & completion_func and completion_template. Mentione that these are mutually exclusive. 

the prompt_func should follow a particular syntax and prompt_template should contain the column names which are available in the dataset.

File: QEfficient/finetune/experimental/configs/README.md [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+*   **completion\_template**: `{output}` → string pattern that tells the fine-tuning pipeline which part of the dataset should be treated as the target output (completion) for the model to learn.
+
+### Example Dataset Configs
+
+### **1. Alpaca (yahma/alpaca-cleaned)**
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/README.md]: These are good examples. Thanks.

File: QEfficient/finetune/experimental/configs/README.md [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+## Overview
+
+This configuration file defines the setup for fine-tuning a Hugging Face causal language model using **LoRA (Low-Rank Adaptation)** and **PEFT (Parameter-Efficient Fine-Tuning)** techniques. It also includes dataset, training, optimizer, and scheduler settings.
+
+***
+### 1. Model Configuration
+
+Model-related parameters for loading and fine-tuning.
+
+*   **model\_type**: `hf` → Type of model (`hf` for Hugging Face, `custom` for custom models).
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/README.md]: Use "hf" to load the model from huggingface. If the user has some custom model then user should inherit from BaseModel class and register the class under a particular key and use the key here.

File: QEfficient/finetune/experimental/configs/README.md [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+*   **auto\_class\_name**: `AutoModelForCausalLM` → AutoClass used to load the model.
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/README.md]: auto_class_name and model_name is applicable in case of "hf" model_type.

File: QEfficient/finetune/experimental/configs/default_config.yaml [hunk 15]
```diff
@@ -33,58 +32,59 @@ dataset:
...
-  completion_only_loss: True
-  report_to: "trackio"
 
-  ddp_config:
-    ddp_backend: "qccl"
-    ddp_find_unused_parameters: False
-    ddp_bucket_cap_mb: 25
-    ddp_broadcast_buffers: null
-    ddp_timeout: 1800
+  # Uncomment if running in Notebook
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/default_config.yaml]: These blocks can be removed if we add the description about these in readme file. For some arguments like FSDP, DeepSpeed we have not finalized how we will use the values. So we will add their documentation in an incremental fashion.

File: QEfficient/finetune/experimental/configs/default_config.yaml [hunk 16]
```diff
@@ -5,21 +5,20 @@
 #
 # -----------------------------------------------------------------------------
 
-# model configuration
+# Model configuration
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/configs/default_config.yaml]: Should we remove the keys and their values if the values are default value based on updated readme? Let me know your thoughts.
